src/emulator/functionsEmulatorRuntime.ts

```diff
@@ -498,99 +497,142 @@ async function getGRPCInsecureCredential(frb: FunctionsRuntimeBundle): Promise<a
 
-  const localAdminModule = require(adminResolution.resolution);
+  // Configuration from the environment
+  const defaultConfig = getDefaultConfig();
+
+  // Configuration for taking to the RTDB emulator
+  const databaseConfig = getDefaultConfig();
+  databaseConfig.databaseURL = `http://localhost:${frb.ports.database}?ns=${frb.projectId}`;
+  databaseConfig.credential = makeFakeCredentials();
 
-  adminModuleProxy = new Proxied<typeof admin>(localAdminModule)
+  const adminModuleProxy = new Proxied<typeof admin>(localAdminModule);
+  const proxiedAdminModule = adminModuleProxy
     .when("initializeApp", (adminModuleTarget) => (opts?: admin.AppOptions, appName?: string) => {
       if (appName) {
         new EmulatorLog("SYSTEM", "non-default-admin-app-used", "", { appName }).log();
         return adminModuleTarget.initializeApp(opts, appName);
+      } else {
+        new EmulatorLog("SYSTEM", "default-admin-app-used", `config=${defaultConfig}`).log();
       }
 
-      const config = JSON.parse(process.env.FIREBASE_CONFIG || "{}");
-      config.credential = makeFakeCredentials();
-
-      if (frb.ports.database) {
-        config.databaseURL = `http://localhost:${frb.ports.database}?ns=${frb.projectId}`;
-        new EmulatorLog("SYSTEM", `Overriding database URL: ${config.databaseURL}`, "").log();
-      }
-
-      const appOptions = {
-        ...config,
+      const defaultAppOptions = {
+        ...defaultConfig,
+        ...opts,
+      };
+      defaultApp = makeProxiedFirebaseApp(frb, adminModuleTarget.initializeApp(defaultAppOptions));
+      logDebug("initializeApp(DEFAULT)", defaultAppOptions);
+
+      // The Realtime Database proxy relies on calling 'initializeApp()' with certain options
+      // (such as credential) that can interfere with other services. Therefore we keep
+      // RTDB isolated in its own FirebaseApp.
+      const databaseAppOptions = {
+        ...databaseConfig,
         ...opts,
       };
-      const originalApp = adminModuleTarget.initializeApp(appOptions);
+      databaseApp = adminModuleTarget.initializeApp(databaseAppOptions, DATABASE_APP);
+      proxiedDatabase = makeProxiedDatabase(adminModuleTarget);
 
-      new EmulatorLog("DEBUG", "default-admin-app-used", "", appOptions).log();
-      logDebug("Intializing default app.", appOptions);
-      defaultApp = proxyFirebaseApp(frb, originalApp, sslCreds);
       return defaultApp;
     })
-    .when("firestore", (adminModuleTarget) => {
-      // When we call admin.firestore() we want to forward this on to the default app,
-      // but if you access something like admin.firestore.FieldValue we don't want to
-      // do anything.
-      const adminFirestoreProxy = new Proxied<typeof admin.firestore>(
-        adminModuleTarget.firestore
-      ).applied(() => {
-        return defaultApp.firestore();
-      });
-
-      return adminFirestoreProxy.finalize();
+    .when("firestore", (target) => {
+      if (frb.ports.firestore) {
```

Comment:
Here's the core of the behavior change: when emulators are running we talk to them.  When they are not, we use prod.

Reply:
This seems to me like duplicate logic to `appProxy` below. Can we reuse the logic, for example, by delegating to `defaultApp.firestore()` and `databaseApp.database()`?

Reply:
The difference is that `admin.firestore` does double duty as a function (`admin.firestore().doc(...)`) and a namespace (`admin.firestore.FieldValue`) whereas the one on the default app does not, so we need this duplication.